sql-plugin/src/main/java/com/nvidia/spark/rapids/UnsafeRowToColumnarBatchIterator.java

```diff
@@ -0,0 +1,150 @@
+
+/**
+ * This class converts UnsafeRow instances to ColumnarBatches on the GPU through the magic of
+ * code generation. This just provides most of the framework a concrete implementation will
+ * be generated based off of the schema.
+ */
+public abstract class UnsafeRowToColumnarBatchIterator implements Iterator<ColumnarBatch> {
+  protected final Iterator<UnsafeRow> input;
+  protected UnsafeRow pending = null;
+  protected final int numRowsEstimate;
+  protected final long dataLength;
+  protected final DType[] rapidsTypes;
+  protected final SQLMetric totalTime;
+  protected final SQLMetric numInputRows;
+  protected final SQLMetric numOutputRows;
+  protected final SQLMetric numOutputBatches;
+
+  protected UnsafeRowToColumnarBatchIterator(
+      Iterator<UnsafeRow> input,
+      Attribute[] schema,
+      CoalesceGoal goal,
+      SQLMetric totalTime,
+      SQLMetric numInputRows,
+      SQLMetric numOutputRows,
+      SQLMetric numOutputBatches) {
+    this.input = input;
+    int sizePerRowEstimate = CudfUnsafeRow.getRowSizeEstimate(schema);
+    numRowsEstimate = (int)Math.max(1,
+        Math.min(Integer.MAX_VALUE - 1, goal.targetSizeBytes() / sizePerRowEstimate));
+    dataLength = ((long) sizePerRowEstimate) * numRowsEstimate;
+    rapidsTypes = new DType[schema.length];
+    for (int i = 0; i < schema.length; i++) {
+      rapidsTypes[i] = GpuColumnVector.getRapidsType(schema[i].dataType());
+    }
+    this.totalTime = totalTime;
+    this.numInputRows = numInputRows;
+    this.numOutputRows = numOutputRows;
+    this.numOutputBatches = numOutputBatches;
+  }
+
+  @Override
+  public boolean hasNext() {
+    return pending != null || input.hasNext();
+  }
+
+  @Override
+  public ColumnarBatch next() {
+    if (!hasNext()) {
+      throw new NoSuchElementException();
+    }
+    final int BYTES_PER_OFFSET = DType.INT32.getSizeInBytes();
+    int dataOffset;
+    int currentRow;
+
+    ColumnVector devColumn;
+    NvtxRange buildRange;
+    try (HostMemoryBuffer dataBuffer = HostMemoryBuffer.allocate(dataLength);
+         HostMemoryBuffer offsetsBuffer =
+             HostMemoryBuffer.allocate(((long)numRowsEstimate + 1) * BYTES_PER_OFFSET)) {
+
+      int[] used = fillBatch(dataBuffer, offsetsBuffer);
+      dataOffset = used[0];
+      currentRow = used[1];
+      // We don't want to loop forever trying to copy nothing
+      assert (currentRow > 0);
+      if (numInputRows != null) {
+        numInputRows.add(currentRow);
+      }
+      if (numOutputRows != null) {
+        numOutputRows.add(currentRow);
+      }
+      if (numOutputBatches != null) {
+        numOutputBatches.add(1);
+      }
+      dataBuffer.incRefCount();
+      offsetsBuffer.incRefCount();
+      try (HostColumnVectorCore dataCv =
+               new HostColumnVectorCore(DType.INT8, dataOffset, Optional.of(0L),
+                   dataBuffer, null, null, new ArrayList<>());
+           HostColumnVector hostColumn = new HostColumnVector(DType.LIST,
```

Comment:
Some comments here as to the high-level flow of things would help, as I was starting to get lost on why a list column vector is showing up here.